pkg/source/reconciler/source/config_watcher.go

```diff
@@ -0,0 +1,160 @@
+	*source.ConfigWatcher
+	logger *zap.SugaredLogger
+
+	kafkaCfg *KafkaConfig
+}
+
+// WatchConfigurations returns a ConfigWatcher initialized with the given
+// options. If no option is passed, the ConfigWatcher observes ConfigMaps for
+// logging, metrics, tracing and Kafka.
+func WatchConfigurations(loggingCtx context.Context, component string,
+	cmw configmap.Watcher) *KafkaSourceConfigWatcher {
+
+	configWatcher := source.WatchConfigurations(loggingCtx, component, cmw)
+
+	cw := &KafkaSourceConfigWatcher{
+		ConfigWatcher: configWatcher,
+		logger:        logging.FromContext(loggingCtx),
+		kafkaCfg:      nil,
+	}
+
+	WithKafka(cw, cmw)
+
+	return cw
+}
+
+// WithKafka observes a Kafka ConfigMap.
+func WithKafka(cw *KafkaSourceConfigWatcher, cmw configmap.Watcher) {
+	cw.kafkaCfg = &KafkaConfig{}
+	watchConfigMap(cmw, KafkaConfigMapName(), cw.updateFromKafkaConfigMap)
+}
+
+func watchConfigMap(cmw configmap.Watcher, cmName string, obs configmap.Observer) {
+	if dcmw, ok := cmw.(configmap.DefaultingWatcher); ok {
+		dcmw.WatchWithDefault(corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{Name: cmName},
+			Data:       map[string]string{},
+		}, obs)
+
+	} else {
+		cmw.Watch(cmName, obs)
+	}
+}
+
+func (cw *KafkaSourceConfigWatcher) updateFromKafkaConfigMap(cfg *corev1.ConfigMap) {
+	kafkaCfg, err := NewKafkaConfigFromConfigMap(cfg)
+	if err != nil {
+		cw.logger.Warnw("ignoring configuration in Kafka configmap ", zap.String("cfg.Name", cfg.Name), zap.Error(err))
+		return
+	}
+
+	cw.kafkaCfg = kafkaCfg
+
+	cw.logger.Debugw("Updated Kafka config from ConfigMap", zap.Any("ConfigMap", cfg))
+	cw.logger.Debugf("Updated Kafka config from ConfigMap 2 %+v", cw.kafkaCfg)
+}
+
+func NewKafkaConfigFromConfigMap(cfg *corev1.ConfigMap) (*KafkaConfig, error) {
+	if cfg == nil {
+		return nil, fmt.Errorf("kafka configmap does not exist")
+	}
+	if _, ok := cfg.Data[constants.SaramaSettingsConfigKey]; !ok {
+		return nil, fmt.Errorf("'%s' key does not exist in Kafka configmap", constants.SaramaSettingsConfigKey)
+	}
+	delete(cfg.Data, "_example")
+	return &KafkaConfig{
+		SaramaYamlString: cfg.Data[constants.SaramaSettingsConfigKey],
+	}, nil
+}
+
+// kafkaConfigEnvVar returns an EnvVar containing the serialized Kafka
+// configuration from the ConfigWatcher.
+func (cw *KafkaSourceConfigWatcher) kafkaConfigEnvVar() corev1.EnvVar {
+	cfg, err := kafkaConfigToJSON(cw.KafkaConfig())
+	if err != nil {
+		cw.logger.Warnw("Error while serializing Kafka config", zap.Error(err))
+	}
+
+	return corev1.EnvVar{
+		Name:  EnvKafkaCfg,
+		Value: cfg,
+	}
+}
+
+func kafkaConfigToJSON(cfg *KafkaConfig) (string, error) { //nolint // for backcompat.
+	if cfg == nil {
+		return "", nil
+	}
+
+	jsonOpts, err := json.Marshal(cfg)
+	return string(jsonOpts), err
+}
+
+// ConfigMapName gets the name of the Kafka ConfigMap
+func KafkaConfigMapName() string {
+	if cm := os.Getenv(kafkaConfigMapNameEnv); cm != "" {
+		return cm
+	}
+	return "config-kafka"
```

Comment:
This is SettingsConfigMapName in pkg/common/constants; should that be used here?